File: backend/python/primeshape/affect.py
import json
from pathlib import Path
from .geometry import signal

RULES = json.loads((Path(__file__).resolve().parents[3] / "config/affect-rules.json").read_text(encoding="utf-8"))


def apparent_affect(blends, quality):
    if quality < 0.4:
        return []
    pair = lambda names: min(blends.get(name, 0) for name in names)
    result = []
    for code, rule in RULES.items():
        if max(blends.get("mouthSmileLeft", 0), blends.get("mouthSmileRight", 0)) >= rule["max_smile"]:
            continue
        if rule.get("max_brow_down") and pair(["browDownLeft", "browDownRight"]) >= rule["max_brow_down"]:
            continue
        values = [max(pair(names) for names in group["features"]) for group in rule["groups"]]
        if all(value >= group["threshold"] for value, group in zip(values, rule["groups"])):
            result.append(signal(code, rule["rotulo"], min(quality, sum(values) / len(values)), "estado_aparente", rule["evidencias"]))
    return result


class AffectTracker:
    def __init__(self):
        self.candidates = {}
        self.last_at = None

    def update(self, signals, now):
        if self.last_at is not None and (now <= self.last_at or now - self.last_at > 0.8):
            self.candidates.clear()
        self.last_at = now
        codes = {item["codigo"] for item in signals}
        self.candidates = {code: value for code, value in self.candidates.items() if code in codes}
        result = []
        for item in signals:
            rule = RULES.get(item["codigo"])
            if rule is None or now - self.candidates.setdefault(item["codigo"], now) >= rule["min_seconds"]:
                result.append(item)
        return result

File: frontend/js/browser/affect.js
import rules from "../../../config/affect-rules.json" with { type:"json" };
import { signal } from "./geometry.js";

export function apparentAffect(blends,quality) {
  if (quality<0.4) return [];
  const pair=names=>Math.min(...names.map(name=>blends[name]||0));
  const result=[];
  for (const [code,rule] of Object.entries(rules)) {
    if (Math.max(blends.mouthSmileLeft||0,blends.mouthSmileRight||0)>=rule.max_smile) continue;
    if (rule.max_brow_down && pair(["browDownLeft","browDownRight"])>=rule.max_brow_down) continue;
    const values=rule.groups.map(group=>Math.max(...group.features.map(pair)));
    if (values.every((value,i)=>value>=rule.groups[i].threshold)) result.push(signal(code,rule.rotulo,Math.min(quality,values.reduce((sum,v)=>sum+v,0)/values.length),"estado_aparente",rule.evidencias));
  }
  return result;
}

export class AffectTracker {
  constructor() { this.reset(); }
  reset() { this.candidates=new Map();this.lastAt=null; }
  update(signals,now) {
    if (this.lastAt!==null && (now<=this.lastAt || now-this.lastAt>0.8)) this.reset();
    this.lastAt=now;
    const codes=new Set(signals.map(s=>s.codigo));
    for (const code of this.candidates.keys()) if (!codes.has(code)) this.candidates.delete(code);
    return signals.filter(s=>{
      const rule=rules[s.codigo];
      if (!rule) return true;
      if (!this.candidates.has(s.codigo)) this.candidates.set(s.codigo,now);
      return now-this.candidates.get(s.codigo)>=rule.min_seconds;
    });
  }
}
